File: src/main/java/de/thm/arsnova/config/WebSocketConfig.java
package de.thm.arsnova.config;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Configuration;
import org.springframework.messaging.simp.config.ChannelRegistration;
import org.springframework.messaging.simp.config.MessageBrokerRegistry;
import org.springframework.web.socket.config.annotation.EnableWebSocketMessageBroker;
import org.springframework.web.socket.config.annotation.StompEndpointRegistry;
import org.springframework.web.socket.config.annotation.WebSocketMessageBrokerConfigurer;

import de.thm.arsnova.config.properties.MessageBrokerProperties;
import de.thm.arsnova.config.properties.SecurityProperties;
import de.thm.arsnova.websocket.handler.AuthChannelInterceptorAdapter;

@Configuration
@EnableWebSocketMessageBroker
@EnableConfigurationProperties(MessageBrokerProperties.class)
public class WebSocketConfig implements WebSocketMessageBrokerConfigurer {
	private static final String MESSAGING_PREFIX = "/backend";
	private static final String[] DESTINATION_PREFIX = {"/exchange", "/topic", "/queue"};
	private static final String USER_REGISTRY_BROADCAST = "/topic/log-user-registry";
	private static final String USER_DESTINATION_BROADCAST = "/queue/log-unresolved-user";

	private final MessageBrokerProperties.Relay relayProperties;
	private final AuthChannelInterceptorAdapter authChannelInterceptorAdapter;
	private String[] corsOrigins;

	@Autowired
	public WebSocketConfig(
			final MessageBrokerProperties messageBrokerProperties,
			final SecurityProperties securityProperties,
			final AuthChannelInterceptorAdapter authChannelInterceptorAdapter) {
		this.relayProperties = messageBrokerProperties.getRelay();
		this.corsOrigins = securityProperties.getCorsOrigins().stream().toArray(String[]::new);
		this.authChannelInterceptorAdapter = authChannelInterceptorAdapter;
	}

	@Override
	public void configureMessageBroker(final MessageBrokerRegistry config) {
		config.setApplicationDestinationPrefixes(MESSAGING_PREFIX);

		if (relayProperties.isEnabled()) {
			config
					.enableStompBrokerRelay(DESTINATION_PREFIX)
					.setUserRegistryBroadcast(USER_REGISTRY_BROADCAST)
					.setUserDestinationBroadcast(USER_DESTINATION_BROADCAST)
					.setRelayHost(relayProperties.getHost())
					.setRelayPort(relayProperties.getPort())
					.setClientLogin(relayProperties.getUsername())
					.setClientPasscode(relayProperties.getPassword());
		} else {
			config.enableSimpleBroker(DESTINATION_PREFIX);
		}
	}

	@Override
	public void registerStompEndpoints(final StompEndpointRegistry registry) {
		registry.addEndpoint("/ws").setAllowedOrigins(corsOrigins).withSockJS();
	}

	@Override
	public void configureClientInboundChannel(final ChannelRegistration registration) {
		registration.interceptors(authChannelInterceptorAdapter);
	}
}

File: websocket/src/main/kotlin/net/particify/arsnova/websocket/config/WebSocketConfig.kt
package net.particify.arsnova.websocket.config

import net.particify.arsnova.websocket.adapter.AuthChannelInterceptorAdapter
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Configuration
import org.springframework.messaging.simp.config.ChannelRegistration
import org.springframework.messaging.simp.config.MessageBrokerRegistry
import org.springframework.scheduling.annotation.EnableScheduling
import org.springframework.web.socket.config.WebSocketMessageBrokerStats
import org.springframework.web.socket.config.annotation.EnableWebSocketMessageBroker
import org.springframework.web.socket.config.annotation.StompEndpointRegistry
import org.springframework.web.socket.config.annotation.WebSocketMessageBrokerConfigurer

@Configuration
@EnableConfigurationProperties(WebSocketProperties::class)
@EnableScheduling
@EnableWebSocketMessageBroker
class WebSocketConfig(
  private val webSocketProperties: WebSocketProperties,
  private val authChannelInterceptorAdapter: AuthChannelInterceptorAdapter,
) : WebSocketMessageBrokerConfigurer {
  private val logger = LoggerFactory.getLogger(javaClass)

  override fun configureMessageBroker(config: MessageBrokerRegistry) {
    config
      .setApplicationDestinationPrefixes(webSocketProperties.messagingPrefix)
      .enableStompBrokerRelay(*webSocketProperties.stomp.destinationPrefix)
      .setUserRegistryBroadcast(webSocketProperties.stomp.userRegistryBroadcast)
      .setUserDestinationBroadcast(webSocketProperties.stomp.userDestinationBroadcast)
      .setRelayHost(webSocketProperties.stomp.relay.host)
      .setRelayPort(webSocketProperties.stomp.relay.port)
      .setClientLogin(webSocketProperties.stomp.relay.user)
      .setClientPasscode(webSocketProperties.stomp.relay.password)
  }

  override fun registerStompEndpoints(registry: StompEndpointRegistry) {
    registry.addEndpoint("/ws").withSockJS()
  }

  override fun configureClientInboundChannel(registration: ChannelRegistration) {
    registration.interceptors(authChannelInterceptorAdapter)
  }

  @Autowired
  fun disableWebSocketStatsLogging(webSocketMessageBrokerStats: WebSocketMessageBrokerStats) {
    webSocketMessageBrokerStats.loggingPeriod = 0
  }
}
